suro-elasticsearch/src/main/java/com/netflix/suro/sink/elasticsearch/ElasticSearchSink.java

```diff
@@ -264,90 +160,199 @@ public String getStat() {
+                    MonitorConfig.builder(PARSING_FAILED)
+                        .withTag(SINK_ID, getSinkId())
+                        .withTag(TagKey.ROUTING_KEY, m.getRoutingKey())
+                        .build()).increment();
+                return null;
+            }
+        }
+    }
+
+    private String getSource(IndexInfo info) throws JsonProcessingException {
+        if (info.getSource() instanceof Map) {
+            return jsonMapper.writeValueAsString(info.getSource());
+        } else {
+            return info.getSource().toString();
         }
     }
 
     @Override
     protected void write(List<Message> msgList) throws IOException {
-        final BulkRequest request = createBulkRequest(msgList);
-
-        senders.execute(createRunnable(request));
+        senders.execute(createRunnable(createBulkRequest(msgList)));
     }
 
     @VisibleForTesting
-    protected BulkRequest createBulkRequest(List<Message> msgList) {
-        final BulkRequest request = new BulkRequest();
+    protected Pair<HttpRequest, List<Message>> createBulkRequest(List<Message> msgList) {
+        List<Message> msgListPayload = new LinkedList<>();
+        StringBuilder sb = new StringBuilder();
         for (Message m : msgList) {
-            IndexRequest indexRequest = createIndexRequest(m);
+            String indexRequest = createIndexRequest(m);
             if (indexRequest != null) {
-                request.add(indexRequest, m);
+                sb.append(indexRequest);
+                msgListPayload.add(m);
             }
         }
-        return request;
+
+        return new Pair<>(
+            HttpRequest.newBuilder()
+                .verb(HttpRequest.Verb.POST)
+                .uri("/_bulk")
+                .setRetriable(true)
+                .entity(sb.toString()).build(),
+            msgListPayload);
     }
 
-    private Runnable createRunnable(final BulkRequest request) {
+    private Runnable createRunnable(final Pair<HttpRequest, List<Message>> request) {
         return new Runnable() {
             @Override
             public void run() {
-                BulkResponse response = client.bulk(request).actionGet();
-                for (BulkItemResponse r : response.getItems()) {
-                    String routingKey = ((Message) request.payloads().get(r.getItemId())).getRoutingKey();
-                    if (r.isFailed() && !r.getFailureMessage().contains("DocumentAlreadyExistsException")) {
-                        log.error("Failed with: " + r.getFailureMessage());
-                        Servo.getCounter(
+                Stopwatch stopwatch = timer.start();
+
+                try {
+                    HttpResponse response = client.executeWithLoadBalancer(request.first());
+                    byte[] bytes = IOUtils.toByteArray(response.getInputStream());
+                    Map<String, Object> result = jsonMapper.readValue(bytes, new TypeReference<Map<String, Object>>(){});
```

Comment:
You don't need to get the bytes. ObjectMapper can read directly from InputStream.